src/openzeppelin/access/accesscontrol/library.cairo

```diff
@@ -10,169 +10,147 @@ from starkware.cairo.common.bool import TRUE, FALSE
-        assert_only_role(admin)
-        _revoke_role(role, user)
-        return ()
-    end
-
-    func renounce_role{
-            syscall_ptr : felt*,
-            pedersen_ptr : HashBuiltin*,
-            range_check_ptr
-        }(role: felt, user: felt):
-        let (caller: felt) = get_caller_address()
-        with_attr error_message("AccessControl: can only renounce roles for self"):
-            assert user = caller
-        end
-        _revoke_role(role, user)
-        return ()
-    end
-
-    #
-    # Unprotected
-    #
-
-    func _grant_role{
-            syscall_ptr : felt*,
-            pedersen_ptr : HashBuiltin*,
-            range_check_ptr
-        }(role: felt, user: felt):
-        let (user_has_role: felt) = has_role(role, user)
-        if user_has_role == FALSE:
-            let (caller: felt) = get_caller_address()
-            AccessControl_role_member.write(role, user, TRUE)
-            RoleGranted.emit(role, user, caller)
-            return ()
-        end
-        return ()
-    end
-
-    func _revoke_role{
-            syscall_ptr : felt*,
-            pedersen_ptr : HashBuiltin*,
-            range_check_ptr
-        }(role: felt, user: felt):
-        let (user_has_role: felt) = has_role(role, user)
-        if user_has_role == TRUE:
-            let (caller: felt) = get_caller_address()
-            AccessControl_role_member.write(role, user, FALSE)
-            RoleRevoked.emit(role, user, caller)
-            return ()
-        end
-        return ()
-    end
-
-    func _set_role_admin{
-            syscall_ptr : felt*,
-            pedersen_ptr : HashBuiltin*,
-            range_check_ptr
-        }(role: felt, admin_role: felt):
-        let (previous_admin_role: felt) = get_role_admin(role)
-        AccessControl_role_admin.write(role, admin_role)
-        RoleAdminChanged.emit(role, previous_admin_role, admin_role)
-        return ()
-    end
-end
+func AccessControl_role_member(role: felt, account: felt) -> (has_role: felt) {
+}
+
+namespace AccessControl {
+    //
+    // Initializer
+    //
+
+    func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
+        ERC165.register_interface(IACCESSCONTROL_ID);
+        return ();
+    }
+
+    //
+    // Modifier
+    //
+
+    func assert_only_role{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        role: felt
+    ) {
+        alloc_locals;
+        let (caller) = get_caller_address();
+        let (authorized) = has_role(role, caller);
+        with_attr error_message("AccessControl: caller is missing role {role}") {
+            assert authorized = TRUE;
+        }
+        return ();
+    }
+
+    //
+    // Getters
+    //
+
+    func has_role{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        role: felt, user: felt
+    ) -> (has_role: felt) {
+        let (authorized: felt) = AccessControl_role_member.read(role, user);
+        return (authorized,);
```

Comment:
It seems that if the return variables don't match, the compiler returns an error when just returning the expression. In other words, this is an issue for external camelCased return values because, for example in AccessControl, the returned `hasRole` from the mock != `has_role` from the lib. Therefore, this won't compile:

```
@view
func hasRole{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
    role: felt, user: felt
) -> (hasRole: felt) {
    return AccessControl.has_role(role, user);
}
```

but this will:

```
@view
func hasRole{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
    role: felt, user: felt
) -> (hasRole: felt) {
    let (hasRole) = AccessControl.has_role(role, user);
    return (hasRole=hasRole);
}
```

and here's the error message of the former:

`
E               starkware.cairo.lang.compiler.preprocessor.preprocessor_error.PreprocessorError: tests/mocks/AccessControl.cairo:22:5: Cannot convert the return type of AccessControl.has_role to the return type of hasRole.
E                   return AccessControl.has_role(role, user);
`

Reply:
but doesn't just returning an unnamed `felt` work? i'm not sure if we want named single return values -- they're tuples. unsure if it has an additional onchain cost, but i'd migrate.

if anything, those are `external` library functions:
> external: subset of public that is ready to be exported as-is by contracts (e.g. transfer on ERC20)

implying they should return the same

Reply:
> but doesn't just returning an unnamed felt on both work?

I initially thought so from the update, but I'm getting this error:

`E               starkware.cairo.lang.compiler.preprocessor.preprocessor_error.PreprocessorError: tests/mocks/AccessControl.cairo:21:6: Only tuple types are supported as the return type of external functions.
E               ) -> felt {`

which is here in StarkWare's repo: https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/starknet/compiler/external_wrapper.py#L679-L698

```
    def process_retdata(
        self,
        func_name: str,
        ret_tuple_type: TypeTuple,
        location: Location,
    ) -> Tuple[Optional[CodeElementFunction], bool]:
        """
        Generates a function that processes the return values. Returns:
        1. The auto-generated function. None if there are no return values.
        2. Whether the ap change is known.
        """

        if len(ret_tuple_type.members) == 0:
            return None, True

        for item in ret_tuple_type.members:
            if item.name is None:
                raise PreprocessorError(
                    "A return value in an external function must be named.", location=item.location
                )
```

Maybe I'm missing something?

Reply:
huh. just asked on telegram, but then we'll have to work around that